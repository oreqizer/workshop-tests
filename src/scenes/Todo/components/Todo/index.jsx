/* @flow */
import * as React from 'react';
import styled from 'styled-components';

import type { Todo } from '../../../../data/Todo';

const Container = styled.div`
  border: 1px solid white;
  height: 65px;
  width: 200px;
  margin: 5px;
  color: white;
`;

const Button = styled.div`
  background: red;
  
  &:hover {
    cursor: pointer;
  }
`;

const Complete = styled.div`
  height: 25px;
  line-height: 25px;
  font-weight: bold;
`;

type Props = {
  todo: Todo,
  onDelete: (id: string) => any,
}

class TodoComponent extends React.PureComponent<Props> {
  props: Props;

  handleDelete() {
    const { todo, onDelete } = this.props;

    onDelete(todo.id);
  }

  render() {
    const { todo } = this.props;

    return (
      <Container>
        <span>
          {todo.text}
        </span>
        <Button onClick={() => this.handleDelete()}>
          DELETE
        </Button>
        <Complete>
          {todo.complete ? 'complete' : 'not complete'}
        </Complete>
      </Container>
    );
  }
}

export default TodoComponent;
