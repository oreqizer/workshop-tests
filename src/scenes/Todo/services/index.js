/* @flow */
import * as R from 'ramda';
import { createSelector } from 'reselect';

import type { Todo } from '../../../data/Todo';
import type { Action } from './actions';
import * as actions from './actions';

export type TodoState = {
  loading: boolean,
  todos: Todo[],
};

const initialState = {
  loading: false,
  todos: [],
};

function todoReducer(state: TodoState = initialState, action: Action) {
  switch (action.type) {
    case actions.GET:
    case actions.CREATE:
    case actions.EDIT:
    case actions.DELETE:
      return R.assoc('loading', true, state);

    case actions.GET_SUCCESS:
      return R.compose(
        R.assoc('todos', action.payload.data),
        R.assoc('loading', false),
      )(state);

    case actions.CREATE_SUCCESS:
      return R.compose(
        R.over(R.lensProp('todos'), R.append(action.payload)),
        R.assoc('loading', false),
      )(state);

    case actions.EDIT_SUCCESS:
      return R.compose(
        R.over(R.lensProp('todos'), R.map(R.ifElse(
          R.propEq('id', action.payload.id),
          R.always(action.payload),
          R.identity,
        ))),
        R.assoc('loading', false),
      )(state);

    case actions.DELETE_SUCCESS:
      return R.compose(
        R.over(R.lensProp('todos'), R.filter(R.compose(
          R.not,
          R.propEq('id', action.payload.id),
        ))),
        R.assoc('loading', false),
      )(state);

    default:
      return state;
  }
}

export default todoReducer;

const getState = state => state.todos;

export const getTodos = createSelector(
  getState, R.prop('todos'),
);

export const getTodosCount = createSelector(
  getTodos, R.length,
);
