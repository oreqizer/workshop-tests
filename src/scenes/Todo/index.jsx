/* @flow */
import * as React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';

import TodoComponent from './components/Todo';
import type { Todo } from '../../data/Todo';
import * as selectors from './services';
import * as actions from './services/actions';
import log from './services/log';

const Container = styled.div`
  width: 100%;
  height: 500px;
  color: white;
`;

const Count = styled.h1`
  color: white;
`;

const Input = styled.input`
  border: 1px solid lime;
`;

const AddButton = styled.button`
  background: lime;
`;

type Props = {
  count: number, // selector
  todos: Todo[], // selector
  getTodos: typeof actions.getTodos,
  createTodo: typeof actions.createTodo,
  deleteTodo: typeof actions.deleteTodo,
  // DI
  log: typeof log,
}

type State = {
  text: string,
}

class TodoContainer extends React.Component<Props, State> {
  static defaultProps = {
    log,
  };

  constructor(props: Props) {
    super(props);

    this.state = { text: '' };
  }

  componentDidMount() {
    this.props.getTodos();
  }

  handleTextChange(ev) {
    this.setState({ text: ev.target.value });
  }

  handleTextKey(ev) {
    if (ev.key === 'Enter') {
      this.handleAdd();
    } else {
      // NOT COVERED
    }
  }

  handleAdd() {
    const { log, createTodo } = this.props;
    const { text } = this.state;

    log('Added new todo with text', text);
    createTodo(text);
    this.setState({ text: '' });
  }

  render() {
    const { count, todos, deleteTodo } = this.props;

    return (
      <Container>
        <Count>
          {count} Todos
        </Count>
        <Input
          id="todo"
          value={this.state.text}
          onKeyDown={ev => this.handleTextKey(ev)}
          onChange={ev => this.handleTextChange(ev)}
          type="text"
        />
        <AddButton onClick={() => this.handleAdd()}>
          ADD
        </AddButton>
        {todos.map(todo => (
          <TodoComponent
            key={todo.id}
            todo={todo}
            onDelete={deleteTodo}
          />
        ))}
      </Container>
    );
  }
}

export default connect(state => ({
  count: selectors.getTodosCount(state),
  todos: selectors.getTodos(state),
}), {
  getTodos: actions.getTodos,
  createTodo: actions.createTodo,
  deleteTodo: actions.deleteTodo,
})(TodoContainer);
