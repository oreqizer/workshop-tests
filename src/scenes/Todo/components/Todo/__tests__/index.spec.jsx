/* @flow */
import * as React from 'react';
import { shallow, mount } from 'enzyme';

import Todo from '../';

const todo = {
  id: '1234',
  text: 'kek',
  complete: false,
};

const todoComplete = {
  id: '1234',
  text: 'kek',
  complete: true,
};

describe('#Todo/index', () => {
  test("snapshot not complete", () => {
    const wrapper = mount(
      <Todo
        todo={todo}
        onDelete={jest.fn()}
      />
    );

    expect(wrapper.getElement()).toMatchSnapshot();
  });

  test("snapshot complete", () => {
    const wrapper = shallow(
      <Todo
        todo={todoComplete}
        onDelete={jest.fn()}
      />
    );

    expect(wrapper.getElement()).toMatchSnapshot();
  });

  test("button onclick", () => {
    const onDelete = jest.fn();
    const wrapper = shallow(
      <Todo
        todo={todo}
        onDelete={onDelete}
      />
    );

    const btn = wrapper.find("[data-test='btn']");

    btn.simulate("click");

    expect(onDelete).toBeCalledWith(todo.id);
  });

  test("instance callback", () => {
    const onDelete = jest.fn();
    const wrapper = shallow(
      <Todo
        todo={todo}
        onDelete={onDelete}
      />
    );

    const node = wrapper.instance();

    node.handleDelete();

    expect(onDelete).toBeCalledWith(todo.id);
  });
});
