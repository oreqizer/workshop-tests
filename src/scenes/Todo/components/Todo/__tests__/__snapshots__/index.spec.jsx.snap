// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Todo/index snapshot complete 1`] = `
<styled.div>
  <span>
    kek
  </span>
  <styled.button
    data-test="btn"
    onClick={[Function]}
  >
    DELETE
  </styled.button>
  <styled.div>
    complete
  </styled.div>
</styled.div>
`;

exports[`#Todo/index snapshot not complete 1`] = `
<TodoComponent
  onDelete={[Function]}
  todo={
    Object {
      "complete": false,
      "id": "1234",
      "text": "kek",
    }
  }
/>
`;
